kotlin/src/com/here/flexiblepolyline/FlexiblePolylineTest.kt

```diff
@@ -0,0 +1,380 @@
+                var z = 0.0
+                if (hasThirdDimension) {
+                    z = coordinates[itr++].trim().replace(")", "").toDouble()
+                }
+                latLngZs.add(FlexiblePolyline.LatLngZ(lat, lng, z))
+            }
+            return latLngZs
+        }
+
+        private fun isNullOrEmpty(str: String?): Boolean {
+            return str == null || str.trim().isEmpty()
+        }
+
+        private fun assertEquals(lhs: Any, rhs: Any?) {
+            if (lhs !== rhs) {
+                if (lhs != rhs) {
+                    throw RuntimeException("Assert failed, $lhs != $rhs")
+                }
+            }
+        }
+
+        private fun assertTrue(value: Boolean) {
+            if (!value) {
+                throw RuntimeException("Assert failed")
+            }
+        }
+
+        private fun <T : Throwable?> assertThrows(expectedType: Class<T>, runnable: Runnable) {
+            try {
+                runnable.run()
+            } catch (actualException: Throwable) {
+                if (!expectedType.isInstance(actualException)) {
+                    println("Working Directory = " + actualException.javaClass.name + " "+ actualException);
+
+                    throw RuntimeException("Assert failed, Invalid exception found!")
+                }
+                return
+            }
+            throw RuntimeException("Assert failed, No exception found!")
+        }
+
+        @JvmStatic
+        fun main(args: Array<String>) {
+            println("Working Directory = " + System.getProperty("user.dir"));
+
+            val DEFAULT_LINE_LENGTH = 1000
+            var lineLength = DEFAULT_LINE_LENGTH
+            if (args.size > 0) {
+                lineLength = args[0].toInt()
+            }
+            val test = FlexiblePolylineTest()
+            test.testInvalidCoordinates()
+            test.testInvalidThirdDimension()
+            test.testConvertValue()
+            test.testSimpleLatLngEncoding()
+            test.testComplexLatLngEncoding()
+            test.testLatLngZEncode()
+            test.encodingSmokeTest()
+
+            //Decode test
+            test.testInvalidEncoderInput()
+            test.testThirdDimension()
+            test.testDecodeConvertValue()
+            test.testSimpleLatLngDecoding()
+            test.testComplexLatLngDecoding()
+            test.testLatLngZDecode()
+            test.decodingSmokeTest()
+            test.testVeryLongLine(lineLength)
+        }
+    }
+}
+
+private data class TestData(
+    val precision: Int = 0,
+    val thirdDimensionPrecision: Int = 0,
+    val thirdDimension: FlexiblePolyline.ThirdDimension? = null,
+    val latLngZs: List<FlexiblePolyline.LatLngZ>? = null
+){}
+
+private data class TestCase(
+    val testInput: String,
+    val testResult: String
+){}
+
+private class TestCaseReader(testInputFile: String, testResultFile: String) : Iterator<TestCase> {
```

Comment:
What about:
```suggestion
private class TestCaseReader(testInputFile: String, testResultFile: String) {
    private val testCases: List<TestCase> =
        Files.readAllLines(Paths.get(FlexiblePolylineTest.TEST_FILES_RELATIVE_PATH).resolve(testInputFile))
        .zip(Files.readAllLines(Paths.get(FlexiblePolylineTest.TEST_FILES_RELATIVE_PATH).resolve(testResultFile)))
        .map { pair ->
            val regex = "\\s|\\(|\\)".toRegex()

            TestCase(
                testInput = pair.first.replace(regex, ""),
                testResult = pair.second.replace(regex, "")
            )
        }

    fun getTestCases(): Iterator<TestCase> = testCases.iterator()
}
```

Reply:
Nice idea, but I usually prefer reading files line by line and avoid reading the full file in memory. Also, I believe the orignal version is a bit easier to understand. So I just changed it to use a list instead of implementing Iterator interface.